@m-kuhn I'm stumped here. The test runs fine locally, but on Travis I get 

```
export LD_LIBRARY_PATH=NOTFOUND:/home/travis/build/qgis/QGIS/build/output/lib:/home/travis/osgeo4travis/lib
export PYTHONPATH=/home/travis/build/qgis/QGIS/build/output/python/:/home/travis/build/qgis/QGIS/build/output/python/plugins:/home/travis/build/qgis/QGIS/tests/src/python:/home/travis/osgeo4travis/lib/python3.3/site-packages/
-- Running /home/travis/virtualenv/python3.3.5/bin/python3 /home/travis/build/qgis/QGIS/python/plugins/processing/tests/QgisAlgorithmsTest.py 
Running from build directory!
- source directory: /home/travis/build/qgis/QGIS
- output directory of the build: /home/travis/build/qgis/QGIS/build/output
!!! prefix path was requested, but it is not valid - we do not run from installed path !!!
..............CMake Error at ProcessingQgisAlgorithmsTest.cmake:15 (MESSAGE):
Traceback (most recent call last):
  File "/home/travis/build/qgis/QGIS/.ci/travis/scripts/ctest2travis.py", line 95, in <module>
    end_fold()
  File "/home/travis/build/qgis/QGIS/.ci/travis/scripts/ctest2travis.py", line 48, in end_fold
    tag = fold_stack.pop()
IndexError: pop from empty list
The command "./.ci/travis/${TRAVIS_CONFIG}/script.sh" exited with 1.
```

With no other useful output or hints why the test is failing... any ideas?

--- a/.ci/travis/scripts/ctest2travis.py
+++ b/.ci/travis/scripts/ctest2travis.py
@@ -45,8 +45,17 @@ def start_fold(tag):
 
 
 def end_fold():
-    tag = fold_stack.pop()
-    sys.stdout.write('travis_fold:end:{}\n'.format(tag))
+    try:
+        tag = fold_stack.pop()
+        sys.stdout.write('travis_fold:end:{}\n'.format(tag))
+    except IndexError:
+        updated_line = colored("======================", 'magenta')
+        updated_line += colored("ctest2travis error when processing the following line:", 'magenta')
+        updated_line += colored("----------------------", 'magenta')
+        updated_line += colored(updated_line, 'magenta')
+        updated_line += colored("----------------------", 'magenta')
+        updated_line += colored("Tried to end fold, but fold was never started.", 'magenta')
+        updated_line += colored("======================", 'magenta')
 
 
 test_count = 0

--- a/python/plugins/processing/tests/testdata/expected/dissolve_field.gml
+++ b/python/plugins/processing/tests/testdata/expected/dissolve_field.gml
@@ -13,7 +13,7 @@
                                                                                                                               
   <gml:featureMember>
     <ogr:dissolve_field fid="polys.1">
-      <ogr:geometryProperty><gml:Polygon srsName="EPSG:4326"><gml:outerBoundaryIs><gml:LinearRing><gml:coordinates>6.241458733205375,-0.054510556621882 7.241458733205375,-1.054510556621882 5.241458733205375,-1.054510556621882 6.241458733205375,-0.054510556621882</gml:coordinates></gml:LinearRing></gml:outerBoundaryIs></gml:Polygon></ogr:geometryProperty>
+      <ogr:geometryProperty><gml:MultiPolygon srsName="EPSG:4326"><gml:polygonMember><gml:Polygon><gml:outerBoundaryIs><gml:LinearRing><gml:coordinates>6.241458733205375,-0.054510556621882 7.241458733205375,-1.054510556621882 5.241458733205375,-1.054510556621882 6.241458733205375,-0.054510556621882</gml:coordinates></gml:LinearRing></gml:outerBoundaryIs></gml:Polygon></gml:polygonMember></gml:MultiPolygon></ogr:geometryProperty>
       <ogr:name>dd</ogr:name>
       <ogr:floatval>0</ogr:floatval>
     </ogr:dissolve_field>
@@ -26,7 +26,7 @@
   </gml:featureMember>
   <gml:featureMember>
     <ogr:dissolve_field fid="polys.0">
-      <ogr:geometryProperty><gml:Polygon srsName="EPSG:4326"><gml:outerBoundaryIs><gml:LinearRing><gml:coordinates>3,2 6,1 6,-3 2,-1 -1,-1 -1,3 3,3 3,2</gml:coordinates></gml:LinearRing></gml:outerBoundaryIs></gml:Polygon></ogr:geometryProperty>
+      <ogr:geometryProperty><gml:MultiPolygon srsName="EPSG:4326"><gml:polygonMember><gml:Polygon><gml:outerBoundaryIs><gml:LinearRing><gml:coordinates>3,2 6,1 6,-3 2,-1 -1,-1 -1,3 3,3 3,2</gml:coordinates></gml:LinearRing></gml:outerBoundaryIs></gml:Polygon></gml:polygonMember></gml:MultiPolygon></ogr:geometryProperty>
       <ogr:name>aa</ogr:name>
       <ogr:intval>1</ogr:intval>
       <ogr:floatval>44.123456</ogr:floatval>
@@ -42,7 +42,7 @@
   </gml:featureMember>
   <gml:featureMember>
     <ogr:dissolve_field fid="polys.7">
-      <ogr:geometryProperty><gml:Polygon srsName="EPSG:4326"><gml:outerBoundaryIs><gml:LinearRing><gml:coordinates>8.162955854126682,2.738771593090211 8.162955854126682,3.738771593090211 9.162955854126682,3.738771593090211 9.162955854126682,2.738771593090211 8.162955854126682,2.738771593090211</gml:coordinates></gml:LinearRing></gml:outerBoundaryIs></gml:Polygon></ogr:geometryProperty>
+      <ogr:geometryProperty><gml:MultiPolygon srsName="EPSG:4326"><gml:polygonMember><gml:Polygon><gml:outerBoundaryIs><gml:LinearRing><gml:coordinates>8.162955854126682,2.738771593090211 8.162955854126682,3.738771593090211 9.162955854126682,3.738771593090211 9.162955854126682,2.738771593090211 8.162955854126682,2.738771593090211</gml:coordinates></gml:LinearRing></gml:outerBoundaryIs></gml:Polygon></gml:polygonMember></gml:MultiPolygon></ogr:geometryProperty>
       <ogr:name>cc</ogr:name>
       <ogr:floatval>0.123</ogr:floatval>
     </ogr:dissolve_field>

--- a/python/plugins/processing/tests/testdata/expected/dissolve_two_fields.gml
+++ b/python/plugins/processing/tests/testdata/expected/dissolve_two_fields.gml
@@ -13,7 +13,7 @@
                                                                                                                               
   <gml:featureMember>
     <ogr:dissolve_two_fields fid="polys.8">
-      <ogr:geometryProperty><gml:Polygon srsName="EPSG:4326"><gml:outerBoundaryIs><gml:LinearRing><gml:coordinates>2.620729366602688,5.088675623800385 2.620729366602688,6.088675623800385 3.620729366602688,6.088675623800385 3.620729366602688,5.088675623800385 2.620729366602688,5.088675623800385</gml:coordinates></gml:LinearRing></gml:outerBoundaryIs></gml:Polygon></ogr:geometryProperty>
+      <ogr:geometryProperty><gml:MultiPolygon srsName="EPSG:4326"><gml:polygonMember><gml:Polygon><gml:outerBoundaryIs><gml:LinearRing><gml:coordinates>2.620729366602688,5.088675623800385 2.620729366602688,6.088675623800385 3.620729366602688,6.088675623800385 3.620729366602688,5.088675623800385 2.620729366602688,5.088675623800385</gml:coordinates></gml:LinearRing></gml:outerBoundaryIs></gml:Polygon></gml:polygonMember></gml:MultiPolygon></ogr:geometryProperty>
       <ogr:name>bb</ogr:name>
       <ogr:intval>2</ogr:intval>
       <ogr:floatval>0.123</ogr:floatval>
@@ -21,7 +21,7 @@
   </gml:featureMember>
   <gml:featureMember>
     <ogr:dissolve_two_fields fid="polys.1">
-      <ogr:geometryProperty><gml:Polygon srsName="EPSG:4326"><gml:outerBoundaryIs><gml:LinearRing><gml:coordinates>6.241458733205375,-0.054510556621882 7.241458733205375,-1.054510556621882 5.241458733205375,-1.054510556621882 6.241458733205375,-0.054510556621882</gml:coordinates></gml:LinearRing></gml:outerBoundaryIs></gml:Polygon></ogr:geometryProperty>
+      <ogr:geometryProperty><gml:MultiPolygon srsName="EPSG:4326"><gml:polygonMember><gml:Polygon><gml:outerBoundaryIs><gml:LinearRing><gml:coordinates>6.241458733205375,-0.054510556621882 7.241458733205375,-1.054510556621882 5.241458733205375,-1.054510556621882 6.241458733205375,-0.054510556621882</gml:coordinates></gml:LinearRing></gml:outerBoundaryIs></gml:Polygon></gml:polygonMember></gml:MultiPolygon></ogr:geometryProperty>
       <ogr:name>dd</ogr:name>
       <ogr:floatval>0</ogr:floatval>
     </ogr:dissolve_two_fields>
@@ -42,7 +42,7 @@
   </gml:featureMember>
   <gml:featureMember>
     <ogr:dissolve_two_fields fid="polys.0">
-      <ogr:geometryProperty><gml:Polygon srsName="EPSG:4326"><gml:outerBoundaryIs><gml:LinearRing><gml:coordinates>3,2 6,1 6,-3 2,-1 -1,-1 -1,3 3,3 3,2</gml:coordinates></gml:LinearRing></gml:outerBoundaryIs></gml:Polygon></ogr:geometryProperty>
+      <ogr:geometryProperty><gml:MultiPolygon srsName="EPSG:4326"><gml:polygonMember><gml:Polygon><gml:outerBoundaryIs><gml:LinearRing><gml:coordinates>3,2 6,1 6,-3 2,-1 -1,-1 -1,3 3,3 3,2</gml:coordinates></gml:LinearRing></gml:outerBoundaryIs></gml:Polygon></gml:polygonMember></gml:MultiPolygon></ogr:geometryProperty>
       <ogr:name>aa</ogr:name>
       <ogr:intval>1</ogr:intval>
       <ogr:floatval>44.123456</ogr:floatval>
@@ -50,7 +50,7 @@
   </gml:featureMember>
   <gml:featureMember>
     <ogr:dissolve_two_fields fid="polys.7">
-      <ogr:geometryProperty><gml:Polygon srsName="EPSG:4326"><gml:outerBoundaryIs><gml:LinearRing><gml:coordinates>8.162955854126682,2.738771593090211 8.162955854126682,3.738771593090211 9.162955854126682,3.738771593090211 9.162955854126682,2.738771593090211 8.162955854126682,2.738771593090211</gml:coordinates></gml:LinearRing></gml:outerBoundaryIs></gml:Polygon></ogr:geometryProperty>
+      <ogr:geometryProperty><gml:MultiPolygon srsName="EPSG:4326"><gml:polygonMember><gml:Polygon><gml:outerBoundaryIs><gml:LinearRing><gml:coordinates>8.162955854126682,2.738771593090211 8.162955854126682,3.738771593090211 9.162955854126682,3.738771593090211 9.162955854126682,2.738771593090211 8.162955854126682,2.738771593090211</gml:coordinates></gml:LinearRing></gml:outerBoundaryIs></gml:Polygon></gml:polygonMember></gml:MultiPolygon></ogr:geometryProperty>
       <ogr:name>cc</ogr:name>
       <ogr:floatval>0.123</ogr:floatval>
     </ogr:dissolve_two_fields>

--- a/python/plugins/processing/tests/testdata/expected/nullGeometryDissolve_output.gml
+++ b/python/plugins/processing/tests/testdata/expected/nullGeometryDissolve_output.gml
@@ -13,7 +13,7 @@
                                                                                                  
   <gml:featureMember>
     <ogr:nullGeometryDissolve_output fid="nullGeometryDissolve.0">
-      <ogr:geometryProperty><gml:Polygon srsName="EPSG:3003"><gml:outerBoundaryIs><gml:LinearRing><gml:coordinates>1756812.050985608249903,5080657.52322755753994 1758058.328737602569163,5080656.446172921918333 1758032.602247689384967,5078868.455123908817768 1756869.935587913962081,5078877.136755040846765 1755007.051325645996258,5078864.273510087281466 1754831.111262238584459,5080811.882167040370405 1756812.050985608249903,5080657.52322755753994</gml:coordinates></gml:LinearRing></gml:outerBoundaryIs></gml:Polygon></ogr:geometryProperty>
+      <ogr:geometryProperty><gml:MultiPolygon srsName="EPSG:3003"><gml:polygonMember><gml:Polygon><gml:outerBoundaryIs><gml:LinearRing><gml:coordinates>1756812.050985608249903,5080657.52322755753994 1758058.328737602569163,5080656.446172921918333 1758032.602247689384967,5078868.455123908817768 1756869.935587913962081,5078877.136755040846765 1755007.051325645996258,5078864.273510087281466 1754831.111262238584459,5080811.882167040370405 1756812.050985608249903,5080657.52322755753994</gml:coordinates></gml:LinearRing></gml:outerBoundaryIs></gml:Polygon></gml:polygonMember></gml:MultiPolygon></ogr:geometryProperty>
     </ogr:nullGeometryDissolve_output>
   </gml:featureMember>
 </ogr:FeatureCollection>

--- a/python/plugins/processing/tests/testdata/qgis_algorithm_tests.yaml
+++ b/python/plugins/processing/tests/testdata/qgis_algorithm_tests.yaml
@@ -252,32 +252,30 @@ tests:
         name: expected/autoincrement_field.gml
         type: vector
 
-#  - algorithm: qgis:dissolve
-#    name: Dissolve using field
-#    params:
-#      DISSOLVE_ALL: false
-#      FIELD: name
-#      INPUT:
-#        name: dissolve_polys.gml
-#        type: vector
-#    results:
-#      OUTPUT:
-#        name: expected/dissolve_field.gml
-#        type: vector
-#
-#  - algorithm: qgis:dissolve
-#    name: Dissolve using two fields
-#    params:
-#      DISSOLVE_ALL: false
-#      FIELD: intval;name
-#      INPUT:
-#        name: dissolve_polys.gml
-#        type: vector
-#    results:
-#      OUTPUT:
-#        name: expected/dissolve_two_fields.gml
-#        type: vector
-#
+  - algorithm: native:dissolve
+    name: Dissolve using field
+    params:
+      FIELD: name
+      INPUT:
+        name: dissolve_polys.gml
+        type: vector
+    results:
+      OUTPUT:
+        name: expected/dissolve_field.gml
+        type: vector
+
+  - algorithm: native:dissolve
+    name: Dissolve using two fields
+    params:
+      FIELD: intval;name
+      INPUT:
+        name: dissolve_polys.gml
+        type: vector
+    results:
+      OUTPUT:
+        name: expected/dissolve_two_fields.gml
+        type: vector
+
 #  - name: Dissolve with geometries reported as valid but as invalid with isGeosValid
 #    algorithm: qgis:dissolve
 #    params:

--- a/src/core/processing/qgsnativealgorithms.cpp
+++ b/src/core/processing/qgsnativealgorithms.cpp
@@ -269,7 +269,7 @@ QString QgsDissolveAlgorithm::shortHelpString() const
   return QObject::tr( "This algorithm takes a polygon or line vector layer and combines their geometries into new geometries. One or more attributes can "
                       "be specified to dissolve only geometries belonging to the same class (having the same value for the specified attributes), alternatively "
                       "all geometries can be dissolved.\n\n"
-                      "If the geometries to be dissolved are spatially separated from each other the output will be multi geometries. "
+                      "All output geometries will be converted to multi geometries. "
                       "In case the input is a polygon layer, common boundaries of adjacent polygons being dissolved will get erased." );
 }
 
@@ -362,26 +362,27 @@ QVariantMap QgsDissolveAlgorithm::processAlgorithm( const QVariantMap &parameter
         break;
       }
 
+      QVariantList indexAttributes;
+      Q_FOREACH ( int index, fieldIndexes )
+      {
+        indexAttributes << f.attribute( index );
+      }
+
+      if ( !attributeHash.contains( indexAttributes ) )
+      {
+        // keep attributes of first feature
+        attributeHash.insert( indexAttributes, f.attributes() );
+      }
+
       if ( f.hasGeometry() && f.geometry() )
       {
-        QVariantList indexAttributes;
-        Q_FOREACH ( int index, fieldIndexes )
-        {
-          indexAttributes << f.attribute( index );
-        }
-
-        if ( !attributeHash.contains( indexAttributes ) )
-        {
-          // keep attributes of first feature
-          attributeHash.insert( indexAttributes, f.attributes() );
-        }
         geometryHash[ indexAttributes ].append( f.geometry() );
       }
     }
 
     int numberFeatures = attributeHash.count();
-    QHash< QVariant, QList< QgsGeometry > >::const_iterator geomIt = geometryHash.constBegin();
-    for ( ; geomIt != geometryHash.constEnd(); ++geomIt )
+    QHash< QVariant, QgsAttributes >::const_iterator attrIt = attributeHash.constBegin();
+    for ( ; attrIt != attributeHash.constEnd(); ++attrIt )
     {
       if ( feedback->isCanceled() )
       {
@@ -389,8 +390,16 @@ QVariantMap QgsDissolveAlgorithm::processAlgorithm( const QVariantMap &parameter
       }
 
       QgsFeature outputFeature;
-      outputFeature.setGeometry( QgsGeometry::unaryUnion( geomIt.value() ) );
-      outputFeature.setAttributes( attributeHash.value( geomIt.key() ) );
+      if ( geometryHash.contains( attrIt.key() ) )
+      {
+        QgsGeometry geom = QgsGeometry::unaryUnion( geometryHash.value( attrIt.key() ) );
+        if ( !geom.isMultipart() )
+        {
+          geom.convertToMultiType();
+        }
+        outputFeature.setGeometry( geom );
+      }
+      outputFeature.setAttributes( attrIt.value() );
       sink->addFeature( outputFeature, QgsFeatureSink::FastInsert );
 
       feedback->setProgress( current * 100.0 / numberFeatures );

--- a/src/core/qgis.cpp
+++ b/src/core/qgis.cpp
@@ -247,9 +247,30 @@ uint qHash( const QVariant &variant )
     case QVariant::Char:
       return qHash( variant.toChar() );
     case QVariant::List:
+
+#if QT_VERSION >= 0x050600
       return qHash( variant.toList() );
+#else
+      {
+        QVariantList list = variant.toList();
+        if ( list.isEmpty() )
+          return -1;
+        else
+          return qHash( list.at( 0 ) );
+      }
+#endif
     case QVariant::StringList:
+#if QT_VERSION >= 0x050600
       return qHash( variant.toStringList() );
+#else
+      {
+        QStringList list = variant.toStringList();
+        if ( list.isEmpty() )
+          return -1;
+        else
+          return qHash( list.at( 0 ) );
+      }
+#endif
     case QVariant::ByteArray:
       return qHash( variant.toByteArray() );
     case QVariant::Date:

